pkg/pool-manager/pool.go

```diff
@@ -186,31 +194,78 @@ func getNextMac(currentMac net.HardwareAddr) net.HardwareAddr {
+		// in OptOutMode namespaces are managed by default
+		isNamespaceManaged = true
+	} else {
+		return false, fmt.Errorf("opt-mode is not defined: %s", vmOptMode)
+	}
+
 	ns, err := p.kubeClient.CoreV1().Namespaces().Get(context.TODO(), namespaceName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get Namespace %s", namespaceName)
+		return isNamespaceManaged, errors.Wrapf(err, "Failed to get Namespace %s", namespaceName)
 	}
 	namespaceLabelMap := ns.GetLabels()
-	log.V(3).Info("namespaceName Labels", "namespaceName", namespaceName, "Labels", namespaceLabelMap)
+	log.V(1).Info("namespaceName Labels", "vm instance namespaceName", namespaceName, "Labels", namespaceLabelMap)
 
 	mutatingWebhookConfiguration, err := p.kubeClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), mutatingWebhookConfigName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
+		return isNamespaceManaged, errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
 	}
 
 	for _, webhook := range mutatingWebhookConfiguration.Webhooks {
 		if webhook.Name == webhookName {
 			if namespaceLabelSet := labels.Set(namespaceLabelMap); namespaceLabelSet != nil {
-				if webhookNamespaceLabelSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector); err == nil && webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
-					return true, nil
+				webhookNamespaceLabelSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+				if err != nil {
+					return isNamespaceManaged, errors.Wrapf(err, "Failed to set webhook Namespace Label Selector for webhook %s", webhook.Name)
 				}
+
+				log.V(1).Info("webhook NamespaceLabelSelectors", "webhookName", webhookName, "NamespaceLabelSelectors", webhookNamespaceLabelSelector)
+				if vmOptMode == OptInMode && webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
+					// if we are in opt-in mode then we check that the namespace has the including label
+					isNamespaceManaged = true
+				} else if vmOptMode == OptOutMode && !webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
+					// if we are in opt-out mode then we check that the namespace does not have the excluding label
+					isNamespaceManaged = false
+				}
+			}
+		}
+	}
+
+	return isNamespaceManaged, nil
+}
+
+func (p *PoolManager) getOptMode(mutatingWebhookConfigName, webhookName string) (OptMode, error) {
+	mutatingWebhookConfiguration, err := p.kubeClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), mutatingWebhookConfigName, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
+	}
+
+	for _, webhook := range mutatingWebhookConfiguration.Webhooks {
+		if webhook.Name == webhookName {
```

Comment:
Can there be multiple webhooks with the same name?

Reply:
my first instinct was to say no, but then I checked, and it's actually technically possible, as there is no validating-webhook in the mutatingWebhookConfiguration instance that checks for duplicate names in the webhook list. So.. that's good to know =)
However, in our case it's still not possible: KMP's mutatingWebhookConfiguration is not meant for user configuration, and even if someone tries to change it manually, it will be reconciled back by the CNAO operator. So I think it's still valid to treat this field as unique and pick by it.

Reply:
See previous comment, you can use a lookup function here to express there is only one webhook name.
This is the 2nd place you can use it.

Reply:
Done